oauthenticator/globus.py

```diff
@@ -174,44 +185,67 @@ def globus_portal_client(self):
                     'globus.org/app/account',
                 ),
             )
+
+        # Each token should have these attributes. Resource server is optional,
+        # and likely won't be present.
+        token_attrs = ['expires_in', 'resource_server', 'scope',
+                       'token_type', 'refresh_token', 'access_token']
+        # The Auth Token is a bit special, it comes back at the top level with the
+        # id token. The id token has some useful information in it, but nothing that
+        # can't be retrieved with an Auth token.
+        # Repackage the Auth token into a dict that looks like the other tokens
+        auth_token_dict = {attr_name: token_json.get(attr_name) for attr_name in token_attrs}
+        # Make sure only the essentials make it into tokens. Other items, such as 'state' are
+        # not needed after authentication and can be discarded.
+        other_tokens = [{attr_name: token_dict.get(attr_name) for attr_name in token_attrs}
+                        for token_dict in token_json['other_tokens']]
+        tokens = other_tokens + [auth_token_dict]
+        # historically, tokens have been organized by resource server for convenience.
+        # If multiple scopes are requested from the same resource server, they will be
+        # combined into a single token from Globus Auth.
+        by_resource_server = {
+                    token_dict['resource_server']: token_dict
+                    for token_dict in tokens
+                    if token_dict['resource_server'] not in self.exclude_tokens
+                }
         return {
             'name': username,
             'auth_state': {
                 'client_id': self.client_id,
-                'tokens': {
-                    tok: v
-                    for tok, v in tokens.by_resource_server.items()
-                    if tok not in self.exclude_tokens
-                },
+                'tokens': by_resource_server,
```

Comment:
The whole `authenticate()` method is very similar to the `GenericOAuthenticator`, where the only functional difference is the shape of tokens it returns. When the admin sets custom scopes, including scopes for third party resource servers secured with Globus Auth, the access/refresh tokens will come through the `other_tokens` key. If the admin didn't enable `auth_state`, this method behaves much the same as the `GenericOAuthenticator`